via/views/_query_params.py

```diff
@@ -1,37 +1,35 @@
 """Methods and data relating to query parameters."""
 
-from urllib.parse import urlencode
-
-# Client configuration query parameters.
-OPEN_SIDEBAR = "via.open_sidebar"
-CONFIG_FROM_FRAME = "via.request_config_from_frame"
-
-
-def strip_client_query_params(base_url, query_params):
-    """Return ``base_url`` with non-Via params from ``query_params`` appended.
-
-    Return ``base_url`` with all the non-Via query params from ``query_params``
-    appended to it as a query string. Any params in ``query_params`` that are
-    meant for Via (the ``"via.*`` query params) will be ignored and *not*
-    appended to the returned URL.
-
-    :param base_url: the protocol, domain and path, for example: https://thirdparty.url/foo.pdf
-    :type base_url: str
-
-    :param query_params: the query params to be added to base_url
-    :type query_params: dict
-
-    :return: ``base_url`` with the non-Via query params appended
-    :rtype: str
-    """
-    client_params = [OPEN_SIDEBAR, CONFIG_FROM_FRAME]
-    filtered_params = urlencode(
-        {
-            param: value
-            for param, value in query_params.items()
-            if param not in client_params
-        }
-    )
-    if filtered_params:
-        return f"{base_url}?{filtered_params}"
-    return base_url
+from urllib.parse import urlencode, urlparse
+
+from webob.multidict import MultiDict
+
+
+class QueryParams:
+    """Client configuration query parameters and related functions."""
+
+    # pylint: disable=too-few-public-methods
+
+    OPEN_SIDEBAR = "via.open_sidebar"
+    CONFIG_FROM_FRAME = "via.request_config_from_frame"
+    ALL_PARAMS = {OPEN_SIDEBAR, CONFIG_FROM_FRAME}
+
+    @classmethod
+    def build_url(cls, url, query, strip_via_params=False):
+        """Add the query to the url, optionally removing via related params.
+
+        :param url: URL to base the new URL on
+        :param query: Dict of query parameters
+        :param strip_via_params: Enable removal of via params
+        :return: A new URL with the relevant query params added
+        """
+
+        # Coerce any immutable NestedMultiDict's into mutable a MultiDict
+        if strip_via_params:
+            query = MultiDict(query)
+
+            bad_keys = [key for key in query.keys() if key.lower() in cls.ALL_PARAMS]
```

Comment:
Process the bad ones up front so we don't change the thing we are iterating over

Reply:
Could be `via_keys` maybe

Reply:
Yep. That's better. We can't dumbly iterate, as theoretically they could in the list 100 times. With multidict you'd have to pop them 100 times to get rid of them. Edge-casey, but might as well do it right.